### PR TITLE
second tutorial (configuration.md)

### PR DESCRIPTION
**Pull request by joelgrus**
in addition,

* renames `run` to `run.py` (I left `run` there so as not to break anyone's workflow, but it logs a "deprecated" warning, and I plan to get rid of it eventually)
* changes "getting started" tutorial to use installation-independent `python -m allennlp.run`
* exports `main` as `allennlp.commands.main()` so that we can tell people to make their own run scripts by importing and calling it

### CODE REVIEW

File: tutorials/getting_started/configuration.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+layout: tutorial
+title: A Deeper Dive Into Configuring Experiments
+id: configuring-experiments
+---
+
+Now that we know how to train and evaluate models,
+let's take a deeper look at our experiment configuration file,
+[tutorials/getting_started/simple_tagger.json](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/simple_tagger.json).
+
+The configuration is a JSON (or [HOCON](https://github.com/typesafehub/config/blob/master/HOCON.md)) object
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: I find this confusing.  Is it JSON or HOCON?  Perhaps you mean "The configuration is HOCON (but don't worry if you are not familiar with HOCON--JSON is a valid subset of HOCON)"

File: tutorials/getting_started/configuration.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+To this end, you can specify "non padded namespaces".
+The default non-padded namespaces are `("*tags", "*labels")`;
+that is, any namespace that ends with `tags` or `labels`.
+
+It turns out that this tutorial dataset does have a small number of POS tags
+in the validation data that don't show up in the training data. This means
+that we need to override the list of non-padded namespaces:
+
+```js
+  "non_padded_namespaces": [],
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: Are we doing this to get padding, or simply `@@UNKNOWN@@` tokens for certain postags?

File: tutorials/getting_started/configuration.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+and a linear layer to convert the output sequences
+into logits representing the probabilities of predicted tags.
+(The last layer is not configurable and so won't appear in our configuration file.)
+
+### The Text Field Embedder
+
+Let's first look at the text field embedder configuration:
+
+```js
+    "text_field_embedder": {
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: The indentation here is quite weird.

Comment (schmmd) [reply tutorials/getting_started/configuration.md]: Would it help to have comments on these lines?

File: tutorials/getting_started/configuration.md [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+Each entry specifies a
+[`TokenEmbedder`](http://docs.allennlp.org/en/latest/api/allennlp.modules.token_embedders.html?highlight=embedding#allennlp.modules.token_embedders.token_embedder.TokenEmbedder)
+that indicates how to embed
+the tokens in that particular namespace. The `TextFieldEmbedder`'s output is the concatenation
+of these embeddings.
+
+The `"tokens"` namespace (which consists of integer encodings of the lowercased words in the input)
+gets fed into an
+[`Embedding`](http://docs.allennlp.org/en/latest/api/allennlp.modules.token_embedders.html?highlight=embedding#allennlp.modules.token_embedders.embedding.Embedding)
+module that embeds the vocabulary words in a 50-dimensional space.
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: append "(specified by `embedding-dim`)"?

File: tutorials/getting_started/configuration.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+concatenated with a 50-dimensional vector for `"token_characters`";
+that is, a 100-dimensional vector.
+
+### The Seq2SeqEncoder
+
+The output of that embedding is processed by our `Seq2SeqEncoder`:
+
+```js
+    "stacked_encoder": {
+            "type": "lstm",
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: I would only indent by 4, rather than by 8.  You could also remove the initial whitespace.

File: tutorials/getting_started/configuration.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+which embeds the characters in an 8-dimensional space
+and then applies a
+[`CnnEncoder`](http://docs.allennlp.org/en/latest/api/allennlp.modules.seq2vec_encoders.html#allennlp.modules.seq2vec_encoders.cnn_encoder.CnnEncoder)
+that uses 50 filters and so also produces a 50-dimensional output. You can see that this encoder also uses a 20% dropout during training.
+
+The output of this `TextFieldEmbedder` is a 50-dimensional vector for `"tokens"`
+concatenated with a 50-dimensional vector for `"token_characters`";
+that is, a 100-dimensional vector.
+
+### The Seq2SeqEncoder
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: Why is this a `Seq2SeqEncoder`?  The JSON below has nothing about `Seq2Seq` but rather specifies `stacked_encoder`.  I'm guessing that `stacked_encoder = Seq2SeqEncoder` but it's an implciit connection.

File: tutorials/getting_started/configuration.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    "num_epochs": 40,
+    "patience": 10,
+    "cuda_device": -1
+  }
+}
+```
+
+Finally, we'll run the training for 40 epochs;
+we'll stop prematurely if we get no improvement for 10 epochs;
+and we'll train on the CPU.
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: You might add instruction (or point somewhere) for how to specify a GPU, as this field is a bit opaque.

File: tutorials/getting_started/configuration.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+That's our entire experiment configuration. If we want to change our optimizer,
+our batch size, our embedding dimensions, or any other hyperparameters,
+all we need to do is modify this config file and `train` another model.
+
+The training configuration is always saved as part of the model archive,
+which means that you can always see how a saved model was trained.
+
+## What's Next
+
+TODO(joelgrus): next part of the tutorial
```
Comment (schmmd) [inline tutorials/getting_started/configuration.md]: TODO

File: tutorials/getting_started/configuration.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+The second, `"token_characters"`, is a
+[`TokenCharactersIndexer`](http://docs.allennlp.org/en/latest/api/allennlp.data.token_indexers.html#allennlp.data.token_indexers.token_characters_indexer.TokenCharactersIndexer)
+that represents each token as a list of int-encoded characters.
+
+Notice that this gives us two different encodings for each token.
+Each encoding has a "namespace", in this case `"tokens"` and `"token_characters"`.
+The `SequenceLabelField` also has a namespace, `"labels"`.
+
+## non-padded namespaces
```
Comment (matt-gardner) [inline tutorials/getting_started/configuration.md]: I was hoping this would be a topic almost all users could ignore, that just shows up in some deep dive.  It's unfortunate that we need to use it here.  Can you construct the validation dataset such that it's not necessary in this tutorial?

Comment (joelgrus) [reply tutorials/getting_started/configuration.md]: yeah, that's a good idea, I'll just filter out the lines with the stray tags

Comment (joelgrus) [reply tutorials/getting_started/configuration.md]: ok, I fixed the datasets and removed all references to non-padded namespaces from the tutorial and config file

Comment (matt-gardner) [reply tutorials/getting_started/configuration.md]: :+1:
